server.js

```diff
@@ -2,31 +2,182 @@ import express from 'express'
+
+// DEFINING ROUTES (endpoints):
+
+// ROOT-ENDPOINT.
+app.get('/', (req, res) => { // req is incoming(we don't change it), the res is outgoing(we do change it)
+  // NOTE TO SELF: Before the send we can write what ever we want here.
+  // For example:
+  // -Database connections
+  // -Complicated operations
+  // -Data lookup
+  // -Third party API-requests
+  if (!res) {
+    res.status(404).json(ERROR_MESSAGE_DATA_NOT_FOUND)
+  }
+  res.send(myEndpoints(app))
 })
 
+// BOOKS-ENDPOINT(WITH PAGES QUERY-PARAM)
+//Q: What's the difference of writing res.send(something) and res.json(something)? ANSWER: https://medium.com/gist-for-js/use-of-res-json-vs-res-send-vs-res-end-in-express-b50688c0cddf
+//Gets all the books from books.json:
+//http://localhost:8080/books
+app.get('/books', (req, res) => {
+  const { author, title } = req.query
+  const page = req.query.page ?? 0
+  const pageSize = req.query.pageSize ?? 20 //The query param is a string
+  let booksList = booksData
+
+  //Query by author:
+  if (author) {
+    booksList = booksList.filter(book => (
+      book.authors.includes(author)
+      || book.authors.toLowerCase().includes(author)
+    ))
+  }
+
+  // Query by title:
+  if (title) {
+    booksList = booksList.filter(book => (
+      book.title.toString().includes(title)
+      || book.title.toString().toLowerCase().includes(title)
+    ))
+  }
+  // PAGINATION (Query by page and/or page-size):
+  // -Page (minValue 0, default 0): number to indicate what page to return.
+  // -PageSize (default: 20)
+  const totalNumberOfBooks = booksList.length
+  const firstPageIndex = 0
+  const startIndex = page * pageSize
+  const endIndex = startIndex + +pageSize // converts pageSize (string) to a number.
+  const bookPage = booksList.slice(startIndex, endIndex) //Creating pages using slice.
+  const returnObject = {
+    totalNumberOfBooks: totalNumberOfBooks,
+    pageSize: pageSize,
+    page: page,
+    firstPageIndex: firstPageIndex,
+    lastPageIndex: parseInt(booksList.length / pageSize),
+    results: bookPage
+  }
+
+  if (bookPage.length === 0) {
+    res.status(404).json(ERROR_MESSAGE_DATA_NOT_FOUND)
+  } else if (!page || !pageSize) {
+    res.json(booksList)
+  } else {
+    res.json(returnObject)
+  }
+})
+
+// LANGUAGE-ENDPOINT. // According to Maks friday lecture: language should rather be query param??
+// Get books in a specified language (language_code):
+// Existing language-codes: 'eng', 'en-US', 'spa', 'fre', 'en-GB', 'mul', 'ger', 'ara', 'por', 'grc'
+app.get('/books/language/:language', (req, res) => {
+  const language = req.params.language
+  const booksByLanguage = booksData.filter((book) => book.language_code === language)
+  
+  // I used existingLanguages to console.log which unique languages that exists in the data.
+  // Would be nice to show to the user somehow which languages could be shown, but I didn't come that far.
+  // const existingLanguages = [...new Set(booksData.map(x => x.language_code))] 
```

Comment:
Yes, I would expect something like that given the name of the endpoint 😇 Endpoints should be named after what they return and not how they are filtered, so as it is now, this filtering could be a query under your main endpoint:
`/books?language=eng`